src/event.rs

```diff
@@ -886,7 +886,22 @@ pub enum Ime {
     /// Notifies when text should be inserted into the editor widget.
     ///
     /// Right before this event winit will send empty [`Self::Preedit`] event.
-    Commit(String),
+    Commit {
+        content: String,
+        /// If selection is Some, the selection / cursor position should be updated after
+        /// placing the `content`.
+        selection: Option<(usize, usize)>,
+        /// If compose_region is Some, the compose region should be updated after replacing the
+        /// text.
+        compose_region: Option<(usize, usize)>,
+    },
+
+    /// Notifies when the text around the cursor should be deleted.
+    /// If `before_length` and `after_length` are [usize::MAX], the entire text should be deleted.
+    DeleteSurroundingText {
+        before_length: usize,
+        after_length: usize,
+    },
```

Comment:
Now that we're exposing a selection region it seems like we need to clarify what the before/after are relative to.

On Android the before would be relative to the start of the selection and the after is relative to the end of the selection.

The Winit docs don't currently say whether we're following the Android notion that the "cursor" is just a special case of a selection where start == end.

With that model though we also need to consider how we will delete / update the selection region?

With the InputConnection API there is a separate setSelection that would make it possible to set the selection start==end for being able to delete all the text.

I'm currently thinking we might need to add a `SetSelection` event for being able to just update the start/end indices of the selection.

Reply:
>With the InputConnection API there is a separate setSelection that would make it possible to set the selection start==end for being able to delete all the text.

>I'm currently thinking we might need to add a SetSelection event for being able to just update the start/end indices of the selection.


The selection should be mapped to the concept of the cursor, so it's effectively the same as in preedit. The `Commit` could set the cursor as well, but I guess the virtual keyboard could move the cursor on its own? Yeah, it  would be fine to have a separate method to advance the selection like you suggest, but we need to make it general on what selection is and probably encapsulate it into a separate type `ImeSelection` or something.